src/test/java/com/amazon/ion/impl/macro/MacroCompilerTest.kt

```diff
@@ -36,79 +36,89 @@ class MacroCompilerTest {
     private fun testCases() = listOf(
         "(macro identity (x) x)" shouldCompileTo TemplateMacro(
             listOf(Parameter("x", Tagged, ParameterCardinality.ExactlyOne)),
-            listOf(Variable(0)),
+            listOf(VariableRef(0)),
         ),
         "(macro identity (any::x) x)" shouldCompileTo TemplateMacro(
             listOf(Parameter("x", Tagged, ParameterCardinality.ExactlyOne)),
-            listOf(Variable(0)),
+            listOf(VariableRef(0)),
         ),
         "(macro pi () 3.141592653589793)" shouldCompileTo TemplateMacro(
             signature = emptyList(),
             body = listOf(DecimalValue(emptyList(), BigDecimal("3.141592653589793")))
         ),
         "(macro cardinality_test (x?) x)" shouldCompileTo TemplateMacro(
             signature = listOf(Parameter("x", Tagged, ParameterCardinality.ZeroOrOne)),
-            body = listOf(Variable(0))
+            body = listOf(VariableRef(0))
         ),
         "(macro cardinality_test (x!) x)" shouldCompileTo TemplateMacro(
             signature = listOf(Parameter("x", Tagged, ParameterCardinality.ExactlyOne)),
-            body = listOf(Variable(0))
+            body = listOf(VariableRef(0))
         ),
         "(macro cardinality_test (x+) x)" shouldCompileTo TemplateMacro(
             signature = listOf(Parameter("x", Tagged, ParameterCardinality.OneOrMore)),
-            body = listOf(Variable(0))
+            body = listOf(VariableRef(0))
         ),
         "(macro cardinality_test (x*) x)" shouldCompileTo TemplateMacro(
             signature = listOf(Parameter("x", Tagged, ParameterCardinality.ZeroOrMore)),
-            body = listOf(Variable(0))
+            body = listOf(VariableRef(0))
         ),
         // Outer 'values' call allows multiple expressions in the body
         // The second `values` is a macro call that has a single argument: the variable `x`
         // The `literal` call causes the third (inner) `(values x)` to be an uninterpreted s-expression.
         """(macro literal_test (x) (values (values x) (literal (values x))))""" shouldCompileTo TemplateMacro(
             signature = listOf(Parameter("x", Tagged, ParameterCardinality.ExactlyOne)),
             body = listOf(
-                MacroInvocation(ByName("values"), startInclusive = 0, endInclusive = 5),
-                MacroInvocation(ByName("values"), startInclusive = 1, endInclusive = 2),
-                Variable(0),
-                SExpValue(emptyList(), startInclusive = 3, endInclusive = 5),
+                MacroInvocation(ByName("values"), selfIndex = 0, endExclusive = 6),
```

Comment:
🗺️ It's not easily noticeable in the diff (because the file was renamed), but I changed the start and end fields in the list, sexp, etc. expressions. I found that I needed the exclusive end rather than the inclusive end in the macro evaluator, and I also found that the "start" I needed was actually `startInclusive + 1`, so I renamed `startInclusive` to be `selfIndex` (since it is the index of this expression instance), and redefined `startInclusive` to specifically refer to the children of the expression.

 (`ktlint` complained that `Expression` should be in a file named `Expression.kt`, otherwise I probably would have left it to make the diff smaller.)